### DISCUSSION TITLE
MATLAB branch and PRs

### DISCUSSION THREAD

**Discussion by KrisThielemans** [swig/swig#2058] [General]
Hi
I'm trying to get my head round current MATLAB status as I might try to push this forward a bit via a student. The situation seems rather complicated. I thought  I'd try to summarise the situation and ping people that have been involved in the past that I found.

@traversaro and @nunoguedelha created https://github.com/robotology/idyntree/issues/688 with a bit of doc, which was helpful in putting this together.

Here's my impression:

- my fork https://github.com/KrisThielemans/swig/commits/matlab is hopelessly out-of-date but does have 4 commits for the test-suite that are possibly not on @jaeandersson's branch (still to confirm)
- https://github.com/swig/swig/tree/matlab is where @wsfulton merged most things from @jaeandersson
- https://github.com/jaeandersson/swig is 12 commits ahead of that. It has a few PRs, including an obvious one from @CameronDevine and a difficult one from @VladimirIvan.
- https://github.com/robotology-dependencies/swig/commits/matlab advanced a bit on that by merging https://github.com/jaeandersson/swig/pull/93, rebasing https://github.com/swig/swig/pull/781, https://github.com/robotology-dependencies/swig/pull/2 and reverting https://github.com/jaeandersson/swig/pull/92 as this broke things.
- @Alzathar created #1519. which seems to have been a leap forward in SWIG history. However, this PR was closed. Files seem to be on https://github.com/Alzathar/swig. I don't know how much of the https://github.com/robotology-dependencies/swig/commits/matlab it has, but it also reverted https://github.com/jaeandersson/swig/pull/92 

Also adding @jgillis

The question is if anyone has any bandwidth for this, even if it's only to suggest a way forward. @wsfulton suggested in #1519 to let @jaeandersson's branch be the "lead" branch, but it seems @jaeandersson has not been able to devote any time on this, so this is probably not feasible anymore.

Suggestions?

**Comment by VladimirIvan** [comment]
@KrisThielemans my PR is very much stale. I'm no longer using Matlab and I gave up on swig for multiple reasons. Good luck with your project.

**Comment by jaeandersson** [comment]
> The question is if anyone has any bandwidth for this, even if it's only to suggest a way forward. @wsfulton suggested in #1519 to let @jaeandersson's branch be the "lead" branch, but it seems @jaeandersson has not been able to devote any time on this, so this is probably not feasible anymore.

Yeah, this this is probably a fair assessment. My needs (as well as @jgillis's needs, but I'll let him speak for himself) were always centered around getting an MATLAB interface that works well for CasADi (https://github.com/casadi/). Once we got there, which we did, the focus was to get the same module to also work for Octave by making it MATLAB/Octave bilingual. The MATLAB module has been pretty stable for us for a long time, but we are also not using the standard typemaps (we've rewritten all typemaps for our own purposes). Using the same module for Octave has been less stable, but that has mostly had to do with limitations in Octave object oriented programming support, not in the SWIG module itself. With most CasADi users obtaining the library as prebuilt binaries for MATLAB, Octave or Python, the inconvenience of not generating the MATLAB/Octave module using the SWIG mainline isn't all that great. Our users don't care if their prebuilt binaries are created using mainline SWIG or a fork of SWIG. Since @jgillis is to thank for most of the work on compiling and distributing CasADi, he should correct me here if I'm wrong, but I just wanted to provide some background as to why I've not been able to spend more time on this.

I'd be happy to provide feedback, but I cannot commit to take a lead on this. I'm stretched too thin with too many others commitments at the moment (including a toddler at home) and my only motivation to help out at this point is really to see the work completed and give back to the SWIG community, which has been very helpful for me/us. Anyway, @KrisThielemans - if you have a student who is willing and able to take the lead on this, that's fantastic. I think it could be beneficial to both the student itself (lots of potential for better C++/MATLAB interoperability) and for the SWIG community.

**Comment by KrisThielemans** [comment]
some info on @Alzathar's branch
- comparing files at [his current head](https://github.com/Alzathar/swig/commit/3ed6aed85bb3631ae0102f1070278721d5575d5f) with the [SWIG master state the last time he merged](https://github.com/Alzathar/swig/commit/b52af4039856cce3dc924d970572350b85f2652d) shows only differences in MATLAB related files.
- compared to @jaeandersson 's branch
    - There are several improvements in `Examples/matlab`, a tiny fix in Examples/test-suite/matlab/Makefile.in
    - No changes in `Lib/matlab`
    - changes in `Source/Modules/matlab.cxx`:
        - small changes to cope with https://github.com/swig/swig/commit/ae044c1c2f91ee17669f7b8bb78748affc8f2883
        - revert https://github.com/jaeandersson/swig/pull/92

***However*** it seems that all MATLAB files were added in [one commit](https://github.com/Alzathar/swig/commit/3ed6aed85bb3631ae0102f1070278721d5575d5f), squashing all MATLAB update history, 

Sadly, to me this makes continuing from  https://github.com/Alzathar/swig history impossible. It implies we'll have to port some of the real changes in the files manually.

**Comment by KrisThielemans** [comment]
info on my branch. This contains the following (I believe valuable) commits that are nowhere else
- https://github.com/KrisThielemans/swig/commit/05e71327268732ac7164cf09be2686db2240bddf
- https://github.com/KrisThielemans/swig/commit/f0b4c49f798b619bf30e4bcd36d81de45c44a38f
- https://github.com/KrisThielemans/swig/commit/9ce53c42e043d11c6ca2ec6049fc8b5ee37a08a4

**Reply by KrisThielemans** [reply]
@jaeandersson would you be able to have a quick look at the above commits and see if you're happy with them on a general level?

**Reply by jaeandersson** [reply]
I see no issues, especially since it's all test-suite/build stuff.

**Comment by KrisThielemans** [comment]
Given all this, and given that `rebase` and related squashing really should ***not*** be done in the context of multiple forks/branches/contributors, I think the best way forward is to
1. take @jaeandersson 's branch and revert https://github.com/jaeandersson/swig/pull/92 as this causes trouble for everyone
2. merge swig/master up to last commit that @Alzathar merged
3. do a `diff` with https://github.com/Alzathar/swig, which should isolate @Alzathar real changes, and create a patch from that
4. merge https://github.com/KrisThielemans/swig/tree/matlab
5. merge https://github.com/robotology-dependencies/swig/pull/2
6. merge https://github.com/jaeandersson/swig/pull/94 if that change hasn't appeared yet
7. merge swig/master up to current commit
8. report on test-suite status
9. do some fixes

Anyone any objections or better ideas?

**Comment by traversaro** [comment]
Thanks a lot to everyone working on this, I can't contribute a lot to the actual development, but I would be happy to provide help as far as I can. Regarding the modification that we have in the fork https://github.com/robotology-dependencies/swig and that we currently use, those are (note that some of this modification may not be desirable to be upstream, but I report them anyhow for alignment): 
* Revert of https://github.com/jaeandersson/swig/pull/92  (see https://github.com/robotology-dependencies/swig/issues/3)
* Another revert of a modification that remove the `subsref` and `subsasn` methods (https://github.com/robotology-dependencies/swig/pull/2, https://github.com/robotology/idyntree/issues/448)
* A backport of https://github.com/swig/swig/pull/781 ( https://github.com/robotology-dependencies/swig/pull/1), that however should be already included if one updates SWIG to 4.0 version 
* Not directly included in the fork, but a hack we did downstream in some package that use SWIG-MATLAB is to rename the `SwigRef` file to avoid conflicts with other packages that install a file with the same name (such as CasADi) but with a content different from the one that we want due to the fact that our version of SWIG diverged due to the  `subsref` and `subsasn` methods (https://github.com/robotology/idyntree/issues/865, https://github.com/robotology/idyntree/pull/868)

**Comment by KrisThielemans** [comment]
ok. the next question is where we put all this. Possibly it makes most sense to do this via PR's to https://github.com/jaeandersson/swig (given history and the fact that there are still some issues listed there), but  it could be easier that I take care of it, and some of you send me PRs. @jaeandersson would do you prefer?

**Reply by jaeandersson** [reply]
@KrisThielemans If you could take care of merging the PRs, that'd be great. If you want, I can make you a collaborator of https://github.com/jaeandersson/swig. But it might be better if you use your own fork, in which case I can still pull those changes to https://github.com/jaeandersson/swig so that they stay in sync.

**Reply by KrisThielemans** [reply]
ok. let me merge them then and give you a humongous PR

**Reply by KrisThielemans** [reply]
See https://github.com/jaeandersson/swig/pull/96

**Comment by KrisThielemans** [comment]
Current state can be seen via https://github.com/swig/swig/compare/master...KrisThielemans:matlab-update

**Comment by KrisThielemans** [comment]
status update: all this has happened in https://github.com/jaeandersson/swig/pull/96. In addition, I've copied the Octave chapter of the manual and adjusted to the best of my knowledge.

Once https://github.com/jaeandersson/swig/pull/96 is merged and GHA is added (end of Jan 2022?), Of course, there are remaining issues but I believe that this qualifies all requirements of an "experimental language" support, (or even more). I therefore that we should do then a PR to mainline SWIG. @wsfulton let us know if this fits your plans.

**Comment by wsfulton** [comment]
Plans are fluid! When it is ready, let us know.
